File: backend/src/main/kotlin/top/bearingwall/asya/service/NotificationService.kt
package top.bearingwall.asya.service

import org.slf4j.LoggerFactory
import org.springframework.messaging.simp.SimpMessagingTemplate
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import top.bearingwall.asya.dto.NotificationEventResponse
import top.bearingwall.asya.dto.NotificationKind
import top.bearingwall.asya.dto.NotificationSyncRequest
import top.bearingwall.asya.model.Instruction
import top.bearingwall.asya.model.InstructionStatus
import top.bearingwall.asya.model.Message
import top.bearingwall.asya.model.MessageReceiver
import top.bearingwall.asya.model.User
import top.bearingwall.asya.model.UserRole
import top.bearingwall.asya.repository.InstructionRepository
import top.bearingwall.asya.repository.MessageReceiverRepository
import top.bearingwall.asya.repository.MessageRepository
import top.bearingwall.asya.repository.UserRepository
import java.time.LocalDateTime
import java.time.ZoneOffset
import java.util.UUID

@Service
class NotificationService(
    private val simpMessagingTemplate: SimpMessagingTemplate,
    private val userRepository: UserRepository,
    private val messageRepository: MessageRepository,
    private val messageReceiverRepository: MessageReceiverRepository,
    private val instructionRepository: InstructionRepository,
) {
    private val log = LoggerFactory.getLogger(NotificationService::class.java)

    @Transactional(readOnly = true)
    fun syncNotifications(delegateUuid: UUID, request: NotificationSyncRequest) {
        val delegate = userRepository.findById(delegateUuid).orElse(null) ?: return
        if (delegate.role != UserRole.DELEGATE) {
            return
        }

        val conferenceUuid = delegate.conference?.uuid ?: return
        val now = LocalDateTime.now(ZoneOffset.UTC)

        val publicMessages = request.publicMessageAfter?.let {
            messageRepository.findAllPublicMessagesPublishedAfter(conferenceUuid, it)
        } ?: messageRepository.findAllPublicMessages(conferenceUuid)
        publicMessages.forEach { sendToUser(delegateUuid, toPublicMessageEvent(it)) }

        val secretEvents = request.secretMessageAfter?.let {
            messageReceiverRepository.findReadableEventsForUserBetween(delegateUuid, it, now)
        } ?: messageReceiverRepository.findReadableEventsForUserBefore(delegateUuid, now)
        secretEvents.forEach { sendToUser(delegateUuid, toSecretMessageEvent(it)) }

        val feedbackInstructions = request.instructionFeedbackAfter?.let {
            instructionRepository.findFeedbackedInstructionsReviewedAfter(delegateUuid, it)
        } ?: instructionRepository.findFeedbackedInstructions(delegateUuid)
        feedbackInstructions.forEach { sendToUser(delegateUuid, toInstructionFeedbackEvent(it)) }

        log.info(
            "Notification sync dispatched, userUuid={}, publicCount={}, secretCount={}, instructionCount={}",
            delegateUuid,
            publicMessages.size,
            secretEvents.size,
            feedbackInstructions.size,
        )
    }

    @Transactional(readOnly = true)
    fun notifyPublicMessage(message: Message) {
        val conferenceUuid = message.conference?.uuid ?: return
        val delegates = userRepository.findAllByConferenceUuidAndRole(conferenceUuid, UserRole.DELEGATE)
        val event = toPublicMessageEvent(message)
        delegates.forEach { delegate ->
            val delegateUuid = delegate.uuid ?: return@forEach
            sendToUser(delegateUuid, event)
        }
    }

    fun notifySecretMessage(messageReceiver: MessageReceiver) {
        val receiver = messageReceiver.receiver ?: return
        if (receiver.role != UserRole.DELEGATE) {
            return
        }
        val receiverUuid = receiver.uuid ?: return
        sendToUser(receiverUuid, toSecretMessageEvent(messageReceiver))
    }

    fun notifyInstructionFeedback(instruction: Instruction) {
        if (instruction.status != InstructionStatus.FEEDBACKED) {
            return
        }
        val submitter = instruction.submitter
        if (submitter.role != UserRole.DELEGATE) {
            return
        }
        val submitterUuid = submitter.uuid ?: return
        sendToUser(submitterUuid, toInstructionFeedbackEvent(instruction))
    }

    @Transactional(readOnly = true)
    fun notifyReadableSecretMessagesBetween(after: LocalDateTime, before: LocalDateTime) {
        if (!after.isBefore(before)) {
            return
        }
        messageReceiverRepository.findReadableEventsBetween(after, before).forEach(::notifySecretMessage)
    }

    private fun sendToUser(userUuid: UUID, event: NotificationEventResponse) {
        log.info("Dispatch notification, userUuid={}, kind={}, eventId={}", userUuid, event.kind, event.eventId)
        simpMessagingTemplate.convertAndSendToUser(userUuid.toString(), "/queue/notifications", event)
    }

    private fun toPublicMessageEvent(message: Message): NotificationEventResponse {
        val messageUuid = message.uuid?.toString() ?: throw IllegalStateException("Message uuid missing")
        val occurredAt = message.publishRealTime
        return NotificationEventResponse(
            eventId = "public:$messageUuid:$occurredAt",
            kind = NotificationKind.PUBLIC_MESSAGE,
            occurredAt = occurredAt,
            messageUuid = messageUuid,
            title = message.title?.trim().orEmpty().ifBlank { "新公开消息" },
            brief = message.brief?.trim().orEmpty().ifBlank { summarizeMessageContent(message.content) },
            senderName = displayName(message.sender),
        )
    }

    private fun toSecretMessageEvent(messageReceiver: MessageReceiver): NotificationEventResponse {
        val message = messageReceiver.message ?: throw IllegalStateException("Message missing")
        val receiverUuid = messageReceiver.receiver?.uuid?.toString()
            ?: throw IllegalStateException("Receiver uuid missing")
        val messageUuid = message.uuid?.toString() ?: throw IllegalStateException("Message uuid missing")
        val occurredAt = messageReceiver.readableAt
        return NotificationEventResponse(
            eventId = "secret:$messageUuid:$receiverUuid:$occurredAt",
            kind = NotificationKind.SECRET_MESSAGE,
            occurredAt = occurredAt,
            messageUuid = messageUuid,
            title = message.title?.trim().orEmpty().ifBlank { "新的非对称消息" },
            brief = message.brief?.trim().orEmpty().ifBlank { summarizeMessageContent(message.content) },
            senderName = displayName(message.sender),
        )
    }

    private fun toInstructionFeedbackEvent(instruction: Instruction): NotificationEventResponse {
        val instructionUuid = instruction.uuid?.toString() ?: throw IllegalStateException("Instruction uuid missing")
        val occurredAt = instruction.reviewedRealTime ?: throw IllegalStateException("Instruction reviewedRealTime missing")
        return NotificationEventResponse(
            eventId = "instruction:$instructionUuid:$occurredAt",
            kind = NotificationKind.INSTRUCTION_FEEDBACK,
            occurredAt = occurredAt,
            instructionUuid = instructionUuid,
            title = instruction.title.ifBlank { "指令已反馈" },
            brief = instruction.reviewComment?.trim().orEmpty().ifBlank { "您的指令已收到反馈，点击查看详情。" },
            senderName = displayName(instruction.reviewedBy),
        )
    }

    private fun displayName(user: User?): String? {
        return user?.displayName?.trim()?.takeIf { it.isNotEmpty() } ?: user?.name
    }

    private fun summarizeMessageContent(content: String?): String {
        val normalized = content?.trim().orEmpty()
        if (normalized.isEmpty()) {
            return "点击查看详情。"
        }
        return if (normalized.length > 64) "${normalized.take(64)}..." else normalized
    }
}

File: backend-dotnet/src/AsyaMun.Api/Services/NotificationService.cs
using System.Globalization;
using AsyaMun.Api.Data;
using AsyaMun.Api.Dtos;
using AsyaMun.Api.Hubs;
using AsyaMun.Api.Models;
using Microsoft.AspNetCore.SignalR;
using Microsoft.EntityFrameworkCore;

namespace AsyaMun.Api.Services;

public class NotificationService
{
    private const int BriefPreviewMaxLength = 64;

    private readonly AppDbContext _db;
    private readonly IHubContext<NotificationHub> _hub;
    private readonly ILogger<NotificationService> _logger;

    public NotificationService(
        AppDbContext db,
        IHubContext<NotificationHub> hub,
        ILogger<NotificationService> logger)
    {
        _db = db;
        _hub = hub;
        _logger = logger;
    }

    public async Task NotifyPublicMessageAsync(Message message)
    {
        try
        {
            var delegateIds = await _db.Users.AsNoTracking()
                .Where(u => u.ConferenceId == message.ConferenceId && u.Role == UserRole.DELEGATE)
                .Select(u => u.Uuid.ToString("D"))
                .ToListAsync();
            if (delegateIds.Count == 0)
            {
                return;
            }

            var e = ToPublicMessageEvent(message);
            await _hub.Clients.Users(delegateIds).SendAsync("Notification", e);
            _logger.LogInformation(
                "Dispatch notification, kind={Kind}, eventId={EventId}, delegateCount={Count}",
                e.Kind,
                e.EventId,
                delegateIds.Count);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to dispatch public message notification, messageUuid={MessageUuid}", message.Uuid);
        }
    }

    public async Task NotifyReadableSecretReceiversAsync(Message message)
    {
        var now = NowUtc();
        foreach (var receiver in message.MessageReceivers)
        {
            if (receiver.ReadableAt <= now)
            {
                await NotifySecretMessageAsync(message, receiver.UserId, receiver.ReadableAt);
            }
        }
    }

    public async Task NotifySecretMessageAsync(Message message, User receiver, DateTime readableAt)
    {
        if (receiver.Role != UserRole.DELEGATE)
        {
            return;
        }

        await SendSecretMessageAsync(message, receiver.Uuid, readableAt);
    }

    public async Task NotifyInstructionFeedbackAsync(Instruction instruction)
    {
        try
        {
            if (instruction.Status != InstructionStatus.FEEDBACKED)
            {
                return;
            }

            var submitter = instruction.Submitter;
            if (submitter == null || submitter.Role != UserRole.DELEGATE)
            {
                return;
            }

            var e = ToInstructionFeedbackEvent(instruction);
            await _hub.Clients.User(submitter.Uuid.ToString("D")).SendAsync("Notification", e);
            _logger.LogInformation(
                "Dispatch notification, kind={Kind}, eventId={EventId}, userUuid={UserUuid}",
                e.Kind,
                e.EventId,
                submitter.Uuid);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to dispatch instruction feedback notification, instructionUuid={InstructionUuid}", instruction.Uuid);
        }
    }

    public async Task NotifyReadableSecretMessagesBetweenAsync(DateTime after, DateTime before)
    {
        if (after >= before)
        {
            return;
        }

        try
        {
            var entries = await _db.MessageReceivers.AsNoTracking()
                .Include(mr => mr.Message)
                .ThenInclude(m => m.Sender)
                .Include(mr => mr.User)
                .Where(mr => mr.Message.IsSecret && mr.ReadableAt > after && mr.ReadableAt <= before)
                .ToListAsync();

            foreach (var entry in entries)
            {
                await NotifySecretMessageAsync(entry.Message, entry.User, entry.ReadableAt);
            }

            if (entries.Count > 0)
            {
                _logger.LogInformation(
                    "Dispatched readable secret message notifications, count={Count}, after={After}, before={Before}",
                    entries.Count,
                    FormatIso(after),
                    FormatIso(before));
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to dispatch readable secret message notifications");
        }
    }

    private async Task NotifySecretMessageAsync(Message message, Guid receiverUuid, DateTime readableAt)
    {
        try
        {
            var isDelegate = await _db.Users.AsNoTracking()
                .AnyAsync(u => u.Uuid == receiverUuid && u.Role == UserRole.DELEGATE);
            if (!isDelegate)
            {
                return;
            }

            await SendSecretMessageAsync(message, receiverUuid, readableAt);
        }
        catch (Exception ex)
        {
            _logger.LogError(
                ex,
                "Failed to dispatch secret message notification, messageUuid={MessageUuid}, receiverUuid={ReceiverUuid}",
                message.Uuid,
                receiverUuid);
        }
    }

    private async Task SendSecretMessageAsync(Message message, Guid receiverUuid, DateTime readableAt)
    {
        var e = ToSecretMessageEvent(message, receiverUuid, readableAt);
        await _hub.Clients.User(receiverUuid.ToString("D")).SendAsync("Notification", e);
        _logger.LogInformation(
            "Dispatch notification, kind={Kind}, eventId={EventId}, userUuid={UserUuid}",
            e.Kind,
            e.EventId,
            receiverUuid);
    }

    private static NotificationEventResponse ToPublicMessageEvent(Message message)
    {
        var messageUuid = message.Uuid.ToString("D");
        var occurredAt = message.PublishRealTime;
        return new NotificationEventResponse(
            $"public:{messageUuid}:{FormatIso(occurredAt)}",
            NotificationKind.PUBLIC_MESSAGE,
            occurredAt,
            messageUuid,
            null,
            string.IsNullOrWhiteSpace(message.Title) ? "新公开消息" : message.Title.Trim(),
            string.IsNullOrWhiteSpace(message.Brief) ? SummarizeMessageContent(message.MsgContent) : message.Brief.Trim(),
            DisplayName(message.Sender));
    }

    private static NotificationEventResponse ToSecretMessageEvent(Message message, Guid receiverUuid, DateTime readableAt)
    {
        var messageUuid = message.Uuid.ToString("D");
        return new NotificationEventResponse(
            $"secret:{messageUuid}:{receiverUuid:D}:{FormatIso(readableAt)}",
            NotificationKind.SECRET_MESSAGE,
            readableAt,
            messageUuid,
            null,
            string.IsNullOrWhiteSpace(message.Title) ? "新的非对称消息" : message.Title.Trim(),
            string.IsNullOrWhiteSpace(message.Brief) ? SummarizeMessageContent(message.MsgContent) : message.Brief.Trim(),
            DisplayName(message.Sender));
    }

    private static NotificationEventResponse ToInstructionFeedbackEvent(Instruction instruction)
    {
        var instructionUuid = instruction.Uuid.ToString("D");
        var occurredAt = instruction.ReviewedRealTime ?? DateTime.SpecifyKind(DateTime.UtcNow, DateTimeKind.Unspecified);
        return new NotificationEventResponse(
            $"instruction:{instructionUuid}:{FormatIso(occurredAt)}",
            NotificationKind.INSTRUCTION_FEEDBACK,
            occurredAt,
            null,
            instructionUuid,
            string.IsNullOrWhiteSpace(instruction.Title) ? "指令已反馈" : instruction.Title.Trim(),
            string.IsNullOrWhiteSpace(instruction.ReviewComment)
                ? "您的指令已收到反馈，点击查看详情。"
                : instruction.ReviewComment.Trim(),
            DisplayName(instruction.ReviewedByNavigation));
    }

    private static string? DisplayName(User? user)
    {
        if (user == null)
        {
            return null;
        }

        return !string.IsNullOrWhiteSpace(user.DisplayName) ? user.DisplayName.Trim() : user.Name;
    }

    private static string SummarizeMessageContent(string? content)
    {
        var normalized = content?.Trim() ?? string.Empty;
        if (normalized.Length == 0)
        {
            return "点击查看详情。";
        }

        return normalized.Length > BriefPreviewMaxLength
            ? $"{normalized[..BriefPreviewMaxLength]}..."
            : normalized;
    }

    private static string FormatIso(DateTime value)
    {
        return value.ToString("yyyy-MM-dd'T'HH:mm:ss", CultureInfo.InvariantCulture);
    }

    private static DateTime NowUtc()
    {
        return DateTime.SpecifyKind(DateTime.UtcNow, DateTimeKind.Unspecified);
    }
}